<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Cohort and Founders · XSim.jl</title><script async src="https://www.googletagmanager.com/gtag/js?id=G-RW1CQJ0L6K"></script><script>  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-RW1CQJ0L6K', {'page_path': location.pathname + location.search + location.hash});
</script><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="XSim.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">XSim.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../../demo/">Demo: Step by Step</a></li><li><span class="tocitem">Core</span><ul><li><a class="tocitem" href="../build_genome/">Build Genome</a></li><li><a class="tocitem" href="../build_phenome/">Build Phenome</a></li><li class="is-active"><a class="tocitem" href>Cohort and Founders</a><ul class="internal"><li class="toplevel"><a class="tocitem" href="#Initialize-a-cohort-by-population-size"><span>Initialize a cohort by population size</span></a></li><li><a class="tocitem" href="#Arguments"><span>Arguments</span></a></li><li><a class="tocitem" href="#Examples"><span>Examples</span></a></li><li class="toplevel"><a class="tocitem" href="#Initialize-a-cohort-by-genotypes/haplotypes-files"><span>Initialize a cohort by genotypes/haplotypes files</span></a></li><li><a class="tocitem" href="#Arguments-2"><span>Arguments</span></a></li><li><a class="tocitem" href="#Example-of-the-demo_genotypes.csv-and-demo_haplotypes.csv"><span>Example of the <code>demo_genotypes.csv</code> and <code>demo_haplotypes.csv</code></span></a></li><li><a class="tocitem" href="#Example"><span>Example</span></a></li><li class="toplevel"><a class="tocitem" href="#Functions-that-insepct-Cohort-properties:"><span>Functions that insepct <code>Cohort</code> properties:</span></a></li><li><a class="tocitem" href="#Genotypes"><span>Genotypes</span></a></li><li><a class="tocitem" href="#QTLs"><span>QTLs</span></a></li><li><a class="tocitem" href="#Breeding-values"><span>Breeding values</span></a></li><li><a class="tocitem" href="#Pedigree"><span>Pedigree</span></a></li><li><a class="tocitem" href="#Minor-Allele-Frequencies-(MAF)"><span>Minor Allele Frequencies (MAF)</span></a></li><li><a class="tocitem" href="#Phenotypes"><span>Phenotypes</span></a></li></ul></li><li><a class="tocitem" href="../mate/">Mating function</a></li><li><a class="tocitem" href="../select/">Selection function</a></li><li><a class="tocitem" href="../breed/">Breed</a></li><li><a class="tocitem" href="../GE/">Genetic Evaluation</a></li></ul></li><li><span class="tocitem">Case Studies</span><ul><li><a class="tocitem" href="../../case/crossbreed/">Rotational Cross-Breeding</a></li><li><a class="tocitem" href="../../case/NAM/">Nested Association Mapping</a></li></ul></li><li><a class="tocitem" href="../../lib/">Library </a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Core</a></li><li class="is-active"><a href>Cohort and Founders</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Cohort and Founders</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/reworkhow/XSim.jl/blob/master/docs/src/core/cohort.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Cohort-and-Founders"><a class="docs-heading-anchor" href="#Cohort-and-Founders">Cohort and Founders</a><a id="Cohort-and-Founders-1"></a><a class="docs-heading-anchor-permalink" href="#Cohort-and-Founders" title="Permalink"></a></h1><div class="admonition is-success"><header class="admonition-header">Cohort and Founders</header><div class="admonition-body"><p>In the current version, <code>Cohort</code> works exactly the same as <code>Founders</code>.</p></div></div><h1 id="Initialize-a-cohort-by-population-size"><a class="docs-heading-anchor" href="#Initialize-a-cohort-by-population-size">Initialize a cohort by population size</a><a id="Initialize-a-cohort-by-population-size-1"></a><a class="docs-heading-anchor-permalink" href="#Initialize-a-cohort-by-population-size" title="Permalink"></a></h1><pre><code class="nohighlight">Cohort(n::Int64=0)</code></pre><h2 id="Arguments"><a class="docs-heading-anchor" href="#Arguments">Arguments</a><a id="Arguments-1"></a><a class="docs-heading-anchor-permalink" href="#Arguments" title="Permalink"></a></h2><ul><li><code>n</code> : An integer to assign the population size.</li></ul><h2 id="Examples"><a class="docs-heading-anchor" href="#Examples">Examples</a><a id="Examples-1"></a><a class="docs-heading-anchor-permalink" href="#Examples" title="Permalink"></a></h2><pre><code class="language-julia-repl">julia&gt; cohort = Cohort(5)
[ Info: Cohort (5 individuals)
[ Info:
[ Info: Mean of breeding values:
[ Info: [1.265 1.697]
[ Info:
[ Info: Variance of breeding values:
[ Info: [1.6 1.4]</code></pre><p>──────────────────────────────────────────────────────────────</p><h1 id="Initialize-a-cohort-by-genotypes/haplotypes-files"><a class="docs-heading-anchor" href="#Initialize-a-cohort-by-genotypes/haplotypes-files">Initialize a cohort by genotypes/haplotypes files</a><a id="Initialize-a-cohort-by-genotypes/haplotypes-files-1"></a><a class="docs-heading-anchor-permalink" href="#Initialize-a-cohort-by-genotypes/haplotypes-files" title="Permalink"></a></h1><pre><code class="nohighlight">Cohort(genetic_data ::Union{DataFrame, Array{Int64}}; args...)
Cohort(filename     ::String; args...)</code></pre><h2 id="Arguments-2"><a class="docs-heading-anchor" href="#Arguments-2">Arguments</a><a class="docs-heading-anchor-permalink" href="#Arguments-2" title="Permalink"></a></h2><ul><li><code>genetic_data</code> : A <code>dataframe</code>/<code>2D-array</code> that stores genotypes/haplotypes in the dimension of individuals by markers.</li><li><code>filename</code> : A <code>filepath</code> to a file storing genotypes/haplotypes data.</li><li><code>n</code> : Number of lines to be loaded from the file. The default value is <code>-1</code> and the entire file will be loaded.</li><li><code>random</code> : By default it&#39;s set to <code>true</code> to randomly select <code>n</code> lines (individuals) from the file to generate the cohort.</li><li><code>alter_maf</code> : It will update MAF based on the provided genotypes if it&#39;s set to <code>true</code> (default).</li></ul><h2 id="Example-of-the-demo_genotypes.csv-and-demo_haplotypes.csv"><a class="docs-heading-anchor" href="#Example-of-the-demo_genotypes.csv-and-demo_haplotypes.csv">Example of the <code>demo_genotypes.csv</code> and <code>demo_haplotypes.csv</code></a><a id="Example-of-the-demo_genotypes.csv-and-demo_haplotypes.csv-1"></a><a class="docs-heading-anchor-permalink" href="#Example-of-the-demo_genotypes.csv-and-demo_haplotypes.csv" title="Permalink"></a></h2><p>Both demo files store marker information for 5 individuals and 4 markers. Use <code>DATA(&quot;demo_genotypes.csv&quot;)</code> to interact with demo files.</p><pre><code class="nohighlight"># demo_genotypes.csv
# rows: individuals, columns: markers
# Homozygote is coded as 0 and 2, otherwise is coded as 1
2,0,0,1
0,0,1,0
0,1,0,2
1,1,0,2
2,0,2,0

# demo_haplotypes.csv
# rows: individuals, columns: alleles
# Reference allele is coded as 0, otherwise is coded as 1
1,1,0,0,0,0,1,0
0,0,0,0,1,0,0,0
0,0,0,1,0,0,1,1
1,0,1,0,0,0,1,1
1,1,0,0,1,1,0,0</code></pre><h2 id="Example"><a class="docs-heading-anchor" href="#Example">Example</a><a id="Example-1"></a><a class="docs-heading-anchor-permalink" href="#Example" title="Permalink"></a></h2><pre><code class="language-julia-repl"># Load entire file
julia&gt; cohort = Cohort(&quot;demo_haplotypes.csv&quot;)
julia&gt; get_genotypes(cohort)
5×4 Array{Int64,2}:
 2  0  0  1
 2  0  2  0
 0  0  1  0
 1  1  0  2
 0  1  0  2

# Randomly load 3 individuals with a dataframe.
julia&gt; data = DATA(&quot;demo_haplotypes.csv&quot;, header=false)
julia&gt; cohort = Cohort(data, random=true, n=3)
julia&gt; get_genotypes(cohort)
3×4 Array{Int64,2}:
 2  0  2  0
 0  1  0  2
 1  1  0  2

# Replace marker MAF by the provided file
julia&gt; cohort = Cohort(&quot;demo_haplotypes.csv&quot;, alter_maf=true)
[ Info: MAF has been updated based on provided haplotypes/genotypes
[ Info: Cohort (5 individuals)
[ Info:
[ Info: Mean of breeding values:
[ Info: [1.418]
[ Info:
[ Info: Variance of breeding values:
[ Info: [2.012]</code></pre><p>──────────────────────────────────────────────────────────────</p><h1 id="Functions-that-insepct-Cohort-properties:"><a class="docs-heading-anchor" href="#Functions-that-insepct-Cohort-properties:">Functions that insepct <code>Cohort</code> properties:</a><a id="Functions-that-insepct-Cohort-properties:-1"></a><a class="docs-heading-anchor-permalink" href="#Functions-that-insepct-Cohort-properties:" title="Permalink"></a></h1><p>All the listed functions can take a keyword argument <code>ID=true</code> to insert individuals&#39; IDs as the first column.</p><h2 id="Genotypes"><a class="docs-heading-anchor" href="#Genotypes">Genotypes</a><a id="Genotypes-1"></a><a class="docs-heading-anchor-permalink" href="#Genotypes" title="Permalink"></a></h2><p>Genotype matirx in the dimension of <code>individuals</code> by <code>markers</code></p><pre><code class="language-julia-repl">julia&gt; get_genotypes(cohort)
5×4 LinearAlgebra.Adjoint{Int64,Array{Int64,2}}:
 0  0  1  0
 2  0  2  0
 2  0  0  1
 0  1  0  2
 1  1  0  2</code></pre><h2 id="QTLs"><a class="docs-heading-anchor" href="#QTLs">QTLs</a><a id="QTLs-1"></a><a class="docs-heading-anchor-permalink" href="#QTLs" title="Permalink"></a></h2><p>QTLs matirx in the dimension of <code>individuals</code> by <code>markers</code></p><pre><code class="language-julia-repl">julia&gt; get_QTLs(cohort)
5×3 Array{Int64,2}:
 2  2  0
 0  0  2
 0  1  0
 1  0  2
 2  0  1</code></pre><h2 id="Breeding-values"><a class="docs-heading-anchor" href="#Breeding-values">Breeding values</a><a id="Breeding-values-1"></a><a class="docs-heading-anchor-permalink" href="#Breeding-values" title="Permalink"></a></h2><p>Breeding values in the dimenstion <code>individuals</code> by <code>traits</code></p><pre><code class="language-julia-repl">julia&gt; get_BVs(cohort)
5×2 LinearAlgebra.Adjoint{Float64,Array{Float64,2}}:
 1.26491   0.0
 3.79473   0.0
 1.26491   1.21268
 0.0       1.69775
 0.632456  1.69775</code></pre><h2 id="Pedigree"><a class="docs-heading-anchor" href="#Pedigree">Pedigree</a><a id="Pedigree-1"></a><a class="docs-heading-anchor-permalink" href="#Pedigree" title="Permalink"></a></h2><p>Pedigree matrix, listed columns are in the order of individuals&#39; ID, sire ID, and dam ID.</p><pre><code class="language-julia-repl">julia&gt; get_pedigree(cohort)
5×3 LinearAlgebra.Adjoint{Int64,Array{Int64,2}}:
1  0  0
2  0  0
3  0  0
4  0  0
5  0  0</code></pre><h2 id="Minor-Allele-Frequencies-(MAF)"><a class="docs-heading-anchor" href="#Minor-Allele-Frequencies-(MAF)">Minor Allele Frequencies (MAF)</a><a id="Minor-Allele-Frequencies-(MAF)-1"></a><a class="docs-heading-anchor-permalink" href="#Minor-Allele-Frequencies-(MAF)" title="Permalink"></a></h2><p>In the case where we have 3 QTLs out of 4 markers, we want to compare their allel frequencies.</p><pre><code class="language-julia-repl">julia&gt; get_MAF(cohort)
4-element Array{Float64,1}:
 0.5
 0.2
 0.3
 0.5</code></pre><h2 id="Phenotypes"><a class="docs-heading-anchor" href="#Phenotypes">Phenotypes</a><a id="Phenotypes-1"></a><a class="docs-heading-anchor-permalink" href="#Phenotypes" title="Permalink"></a></h2><p>Simulate cohort phenotypes based on the defined <code>phenome</code>. <code>h2</code> and <code>ve</code> can be assigned specifically for this one-time simulation.</p><pre><code class="language-julia-repl">julia&gt; get_phenotypes(cohort)
5×1 Array{Float64,2}:
  1.1126064336992942
 -0.8337021175232547
 -0.363621019381922
  4.042256656472762
  1.7828800511223049</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../build_phenome/">« Build Phenome</a><a class="docs-footer-nextpage" href="../mate/">Mating function »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.3 on <span class="colophon-date" title="Friday 23 July 2021 11:30">Friday 23 July 2021</span>. Using Julia version 1.5.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
