<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Select · XSim.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="XSim.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">XSim.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../../demo/">Demo: Step by Step</a></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../build_genome/">Build Genome</a></li><li><a class="tocitem" href="../build_phenome/">Build Phenome</a></li><li><a class="tocitem" href="../cohort/">Cohort and Founders</a></li><li><a class="tocitem" href="../mate/">Mate</a></li><li class="is-active"><a class="tocitem" href>Select</a></li><li><a class="tocitem" href="../breed/">Breed</a></li></ul></li><li><span class="tocitem">Case Studies</span><ul><li><a class="tocitem" href="../../case/crossbreed/">Rotational Cross-Breeding</a></li><li><a class="tocitem" href="../../case/NAM/">NAM</a></li></ul></li><li><span class="tocitem">Genetic Evaluation</span><ul><li><a class="tocitem" href="../../ge/gblup_pblup/">GBLUP and PBLUP Evaluation</a></li><li><a class="tocitem" href="../../ge/multi_trait/">Multi-Trait Genetic Evaluation</a></li></ul></li><li><span class="tocitem">For Developers</span><ul><li><a class="tocitem" href="../../dev/custom_data/">Customized Data</a></li><li><a class="tocitem" href="../../dev/expand_genome/">Expand Genome Information</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Library</a></li><li class="is-active"><a href>Select</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Select</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/reworkhow/XSim.jl/blob/master/docs/src/lib/select.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Select"><a class="docs-heading-anchor" href="#Select">Select</a><a id="Select-1"></a><a class="docs-heading-anchor-permalink" href="#Select" title="Permalink"></a></h1><article class="docstring"><header><a class="docstring-binding" id="Main.XSim.select" href="#Main.XSim.select"><code>Main.XSim.select</code></a> — <span class="docstring-category">Function</span></header><section><div><p><strong>Selection function</strong></p><pre><code class="nohighlight">select(cohort      ::Cohort,
       n           ::Int64,
       criteria    ::String = &quot;phenotypes&quot;;
       h2          ::Union{Array{Float64}, Float64}=GLOBAL(&quot;h2&quot;),
       ve          ::Union{Array{Float64}, Float64}=GLOBAL(&quot;Ve&quot;),
       weights     ::Array{Float64, 1}             =[1.0],
       return_log  ::Bool                          =false,
       is_random   ::Bool                          =false,
       silent      ::Bool                          =GLOBAL(&quot;silent&quot;),
       args...)

select(cohort::Cohort, ratio::Float64; args...)</code></pre><p><strong>Arguments</strong></p><p>Positional arguments</p><ul><li><code>cohort</code> : A <code>cohort</code> from which individuals are selected.</li><li><code>n</code> : <code>n</code> individuals are selected.</li><li><code>ratio</code> : <code>ratio</code> portion of individuals are selected.</li></ul><p>Keyword arguments</p><ul><li><code>criteria</code> : <code>Criteria</code> that will be used for the selecition. Default &quot;phenotypes&quot;, the options are [&quot;phenotypes&quot;, &quot;GBLUP&quot;]. If set to &quot;GBLUP&quot;, a genetic evaluation is carried out by <code>JWAS</code> and the estimated breeding values will be the <code>criteria</code>.</li><li><code>h2</code> : The heritability <code>h2</code> of the simulated phenotypes.</li><li><code>ve</code> : The residual covariance <code>ve</code> of the simulated phenotypes.</li><li><code>weight</code> : Linear coefficients of traits for the selection. The selection is more sensitive to traits with greater <code>weight</code>. Negative</li><li><code>return_log</code> : Default <code>false</code>. Set <code>true</code> to return selection differential and selection response besides the selected cohort.</li><li><code>silent</code> : Default <code>false</code>. Set <code>true</code> to mute the log messages.</li></ul><p><strong>Outputs</strong></p><p>A selected <code>cohort</code> object will be returned. If <code>return_log</code> is set to <code>true</code>, a <code>dictionary</code> object containing the selected cohort, selection differential, and selection response will be returned.</p><p>─────────────────────────────────────────────────────────</p><p><strong>Example</strong></p><p><strong>Single Trait Selection</strong></p><p>Set demo genome and phenome with single traits controlled by 50 QTLs.</p><pre><code class="language-julia-repl">julia&gt; build_demo()
julia&gt; build_phenome(50)
julia&gt; summary()
[ Info: --------- Genome Summary ---------
[ Info: Number of Chromosome  : 10
[ Info: 
[ Info: Chromosome Length (cM): 1500.0
[ Info: [150.0, 150.0, 150.0, 150.0, 150.0, 150.0, 150.0, 150.0, 150.0, 150.0]
[ Info: 
[ Info: Number of Loci        : 1000
[ Info: [100, 100, 100, 100, 100, 100, 100, 100, 100, 100]
[ Info: 
[ Info: Genotyping Error      : 0.0
[ Info: Mutation Rate         : 0.0
[ Info: 
[ Info: --------- Phenome Summary ---------
[ Info: Number of Traits      : 1
[ Info: Heritability (h2)     : [0.5]
┌ Info: 
│   Genetic_Variance =
│    1×1 Array{Float64,2}:
└     1.0
┌ Info: 
│   Residual_Variance =
│    1×1 Array{Float64,2}:
└     1.0
[ Info: Number of QTLs        : [50]</code></pre><p>Initialize a cohort with 100 individuals</p><pre><code class="language-julia-repl">julia&gt; cohort = Cohort(100)</code></pre><p><strong>Select 30 individuals</strong></p><pre><code class="language-julia-repl"># Select top 30 individuals
julia&gt; cohort_s = select(cohort, 30)
# Equivalent
julia&gt; cohort_s = select(cohort, 0.3)

[ Info: --------- Selection Summary ---------
[ Info: Select 30 individuals out of 100 individuals
[ Info: Selection differential (P): [1.174]
[ Info: Selection response     (G): [0.843]
┌ Info:
│   Residual_Variance =
│    1×1 Array{Float64,2}:
└     1.0
[ Info: --------- Offsprings Summary ---------
[ Info: Cohort (30 individuals)
[ Info:
[ Info: Mean of breeding values:
[ Info: [1.448]
[ Info:
[ Info: Variance of breeding values:
[ Info: [0.367]</code></pre><p><strong>Assign Heritability <code>h2</code> or Residual Covariance <code>ve</code></strong></p><pre><code class="language-julia-repl"># Assign heritability
julia&gt; progenies = select(cohort, 30, h2=0.1)

# Equivalent in the case where genetic variance `vg` is 1.0
julia&gt; progenies = select(cohort, 30, ve=9.0)

[ Info: --------- Selection Summary ---------
[ Info: Select 30 individuals out of 100 individuals
[ Info: Selection differential (P): [1.182]
[ Info: Selection response     (G): [0.338]
┌ Info:
│   Residual_Variance =
│    1×1 Array{Float64,2}:
└     9.0
[ Info: --------- Offsprings Summary ---------
[ Info: Cohort (30 individuals)
[ Info: 
[ Info: Mean of breeding values: 
[ Info: [0.956]
[ Info: 
[ Info: Variance of breeding values: 
[ Info: [0.643]</code></pre><p><strong>Negative Selection</strong></p><p>Set <code>is_positive=false</code> to rank individuals in ascending order</p><pre><code class="language-julia-repl">julia&gt; progenies = select(cohort, 30, is_positive=false)
[ Info: --------- Selection Summary ---------
[ Info: Select 30 individuals out of 100 individuals
[ Info: Selection differential (P): [-1.19]
[ Info: Selection response     (G): [-0.89]
┌ Info: 
│   Residual_Variance =
│    1×1 Array{Float64,2}:
└     1.0
[ Info: --------- Offsprings Summary ---------
[ Info: Cohort (30 individuals)
[ Info: 
[ Info: Mean of breeding values: 
[ Info: [-0.24]
[ Info: 
[ Info: Variance of breeding values: 
[ Info: [0.566]</code></pre><p><strong>Random Selection</strong></p><pre><code class="language-julia-repl">julia&gt; progenies = select(cohort, 30, is_random=true)
[ Info: --------- Selection Summary ---------
[ Info: Select 30 individuals out of 100 individuals
[ Info: Selection differential (P): [-0.06]
[ Info: Selection response     (G): [-0.191]
┌ Info: 
│   Residual_Variance =
│    1×1 Array{Float64,2}:
└     1.0
[ Info: --------- Offsprings Summary ---------
[ Info: Cohort (30 individuals)
[ Info: 
[ Info: Mean of breeding values: 
[ Info: [0.441]
[ Info: 
[ Info: Variance of breeding values: 
[ Info: [0.946]</code></pre><p><strong>Selection wiht Multiple Parameters</strong></p><p>It&#39;s possible specify multiple parameters described above in one selection. User can either enclose parameters as keyword arguments, or pass them through a <code>dictionary</code> object.</p><pre><code class="language-julia-repl"># Keyword args
julia&gt; progenies = select(cohort, 30, h2=0.3, is_positive=false)

# Equivalent
julia&gt; args = Dict(:h2=&gt;0.3,
                   :is_positive=&gt;false)
julia&gt; progenies = select(cohort, 30; args...)

[ Info: --------- Selection Summary ---------
[ Info: Select 30 individuals out of 100 individuals
[ Info: Selection differential (P): [-1.086]
[ Info: Selection response     (G): [-0.486]
┌ Info: 
│   Residual_Variance =
│    1×1 Array{Float64,2}:
└     2.3333333333333335
[ Info: --------- Offsprings Summary ---------
[ Info: Cohort (30 individuals)
[ Info: 
[ Info: Mean of breeding values: 
[ Info: [0.154]
[ Info: 
[ Info: Variance of breeding values: 
[ Info: [0.818]</code></pre><p>─────────────────────────────────────────────────────────</p><p><strong>Multi-Trait Selection</strong></p><p>Set demo genome and phenome with single traits controlled by 50 QTLs.</p><pre><code class="language-julia-repl">julia&gt; build_demo()
julia&gt; summary()
[ Info: --------- Genome Summary ---------
[ Info: Number of Chromosome  : 10
[ Info: 
[ Info: Chromosome Length (cM): 1500.0
[ Info: [150.0, 150.0, 150.0, 150.0, 150.0, 150.0, 150.0, 150.0, 150.0, 150.0]
[ Info: 
[ Info: Number of Loci        : 1000
[ Info: [100, 100, 100, 100, 100, 100, 100, 100, 100, 100]
[ Info: 
[ Info: Genotyping Error      : 0.0
[ Info: Mutation Rate         : 0.0
[ Info: 
[ Info: --------- Phenome Summary ---------
[ Info: Number of Traits      : 2
[ Info: Heritability (h2)     : [0.5, 0.5]
┌ Info: 
│   Genetic_Variance =
│    2×2 Array{Float64,2}:
│     1.0  0.0
└     0.0  1.0
┌ Info: 
│   Residual_Variance =
│    2×2 Array{Float64,2}:
│     1.0  0.0
└     0.0  1.0
[ Info: Number of QTLs        : [3 8]</code></pre><p>Initialize a cohort with 100 individuals</p><pre><code class="language-julia-repl">julia&gt; cohort = Cohort(100)</code></pre><p><strong>Assign Heritabilities for Multiple Traits</strong></p><pre><code class="language-julia-repl">julia&gt; progenies = select(cohort, 30, h2=[0.9, 0.3])
[ Info: --------- Selection Summary ---------
[ Info: Select 30 individuals out of 100 individuals
[ Info: Selection differential (P): [0.468 1.028]
[ Info: Selection response     (G): [0.383 0.636]
┌ Info: 
│   Residual_Variance =
│    2×2 Array{Float64,2}:
│     0.111111  0.0
└     0.0       2.33333
[ Info: --------- Offsprings Summary ---------
[ Info: Cohort (30 individuals)
[ Info: 
[ Info: Mean of breeding values: 
[ Info: [-0.889 0.28]
[ Info: 
[ Info: Variance of breeding values: 
[ Info: [0.947 0.625]</code></pre><p><strong>Assign Trait Correlations via Residual Covariance</strong></p><pre><code class="language-julia-repl">julia&gt; progenies = select(cohort, 30, ve=[1   0.3
                                          0.3   1])
[ Info: --------- Selection Summary ---------
[ Info: Select 30 individuals out of 100 individuals
[ Info: Selection differential (P): [0.866 0.925]
[ Info: Selection response     (G): [0.662 0.762]
┌ Info: 
│   Residual_Variance =
│    2×2 Array{Float64,2}:
│     1.0  0.3
└     0.3  1.0
[ Info: --------- Offsprings Summary ---------
[ Info: Cohort (30 individuals)
[ Info: 
[ Info: Mean of breeding values: 
[ Info: [-0.608 0.406]
[ Info: 
[ Info: Variance of breeding values: 
[ Info: [0.549 0.476]</code></pre><p><strong>Derive Selection Index for Multiple Traits</strong></p><p>Assigning a vector to the parameter <code>weights</code> to derive a selection index which is a linear combintation of the weights and the phenotypes.  In this example, we demonstrate two traits with the heritability of 0.3 and 0.8, respectively. And we can select traits with more weight on the second trait which is more heritable, and negatively select the first trait.</p><pre><code class="language-julia-repl">julia&gt; progenies = select(cohort, 30, h2=[.3, .8], weights=[-0.1, 0.9])
[ Info: --------- Selection Summary ---------
[ Info: Select 30 individuals out of 100 individuals
[ Info: Selection differential (P): [-0.318 1.027]
[ Info: Selection response     (G): [-0.233 0.869]
┌ Info: 
│   Residual_Variance =
│    2×2 Array{Float64,2}:
│     2.33333  0.0
└     0.0      0.25
[ Info: --------- Offsprings Summary ---------
[ Info: Cohort (30 individuals)
[ Info: 
[ Info: Mean of breeding values: 
[ Info: [-1.508 0.513]
[ Info: 
[ Info: Variance of breeding values: 
[ Info: [1.053 0.458]</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/reworkhow/XSim.jl/blob/a1b6c8b18c994790a0f7ad915c0bdc61847af075/src/core/select.jl#L1-L314">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../mate/">« Mate</a><a class="docs-footer-nextpage" href="../breed/">Breed »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.3 on <span class="colophon-date" title="Thursday 22 July 2021 16:15">Thursday 22 July 2021</span>. Using Julia version 1.5.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
